sql-plugin/src/main/scala/org/apache/spark/sql/rapids/AggregateFunctions.scala

```diff
@@ -444,33 +444,86 @@ object CudfNthLikeAggregate {
  * In the future, this aggregate class should be removed and the mean values should be
  * generated in the output of libcudf's M2 aggregate.
  */
-class CudfMean(override val dataType: DataType) extends CudfAggregate {
-  override lazy val reductionAggregate: cudf.ColumnVector => cudf.Scalar = _ =>
-    throw new UnsupportedOperationException("CudfMean is not supported in reduction")
+class CudfMean extends CudfAggregate {
+  override lazy val reductionAggregate: cudf.ColumnVector => cudf.Scalar =
+    (col: cudf.ColumnVector) => {
+      val count = col.getRowCount - col.getNullCount
+      if (count == 0) {
+        Scalar.fromDouble(0.0)
+      } else {
+        withResource(col.sum(DType.FLOAT64)) { sum =>
+          Scalar.fromDouble(sum.getDouble / count.toDouble)
+        }
+      }
+    }
 
-  override lazy val groupByAggregate: GroupByAggregation =
-    GroupByAggregation.mean()
+  override lazy val groupByAggregate: GroupByAggregation = GroupByAggregation.mean()
 
   override val name: String = "CudfMeanForM2"
+
+  override def dataType: DataType = DoubleType
 }
 
 class CudfM2 extends CudfAggregate {
-  override lazy val reductionAggregate: cudf.ColumnVector => cudf.Scalar = _ =>
-    throw new UnsupportedOperationException("CudfM2 aggregation is not supported in reduction")
+  override lazy val reductionAggregate: cudf.ColumnVector => cudf.Scalar =
+    (col: cudf.ColumnVector) => {
+      val count = col.getRowCount - col.getNullCount
+      if (count == 0) {
+        Scalar.fromDouble(0.0)
+      } else {
+        withResource(col.sum(DType.FLOAT64)) {sum =>
+          val mean = sum.getDouble / count.toDouble
+          withResource(col.reduce(ReductionAggregation.sumOfSquares(), DType.FLOAT64)) { sumSqr =>
+            Scalar.fromDouble(sumSqr.getDouble - mean * mean * count.toDouble)
+          }
+        }
+      }
+    }
 
-  override lazy val groupByAggregate: GroupByAggregation =
-    GroupByAggregation.M2()
+  override lazy val groupByAggregate: GroupByAggregation = GroupByAggregation.M2()
 
   override val name: String = "CudfM2"
   override def dataType: DataType = DoubleType
 }
 
 class CudfMergeM2 extends CudfAggregate {
-  override lazy val reductionAggregate: cudf.ColumnVector => cudf.Scalar = _ =>
-    throw new UnsupportedOperationException("CudfMergeM2 aggregation is not supported in reduction")
+  override lazy val reductionAggregate: cudf.ColumnVector => cudf.Scalar =
+    (col: cudf.ColumnVector) => {
+      withResource(col.copyToHost()) { hcv =>
```

Comment:
nit: Would be nice to have a NVTX Range here so we can see in traces how long this is taking.